prologin/workernode/operations.py

```diff
@@ -46,308 +50,319 @@ def untar(content, path, compression='gz'):
-    return isolator
-
-
-async def compile_champion(config, ctgz):
-    """
-    Compiles the champion contained in ctgz and returns a tuple TODO
-
-    """
-    ctgz = b64decode(ctgz)
-    code_dir = os.path.abspath(os.path.dirname(__file__))
-    compile_script = os.path.join(code_dir, 'compile-champion.sh')
-
-    limits = {
-        'wall-time': config['timeout'].get('compile', 400),
-        'fsize': 50 * 1024,
-    }
-    allowed_dirs = ['/tmp:rw', code_dir, '/etc', config['path']['player_env']]
-
-    isolator = isolate.Isolator(limits, allowed_dirs=allowed_dirs)
-    async with isolator:
-        compiled_path = isolator.path / 'champion-compiled.tar.gz'
-        log_path = isolator.path / 'compilation.log'
-        with (isolator.path / 'champion.tgz').open('wb') as f:
-            f.write(ctgz)
-
-        cmd = [compile_script, config['path']['player_env'], '/box']
-        await isolator.run(cmd)
-        ret = isolator.isolate_retcode == 0
-
-        compilation_content = b''
-        if ret:
+        async with isolator:
+            self.isolator = isolator
             try:
-                with compiled_path.open('rb') as f:
-                    compilation_content = f.read()
-            except FileNotFoundError:
-                ret = False
-
-        log = ''
+                yield isolator
+            finally:
+                self.result['success'] = isolator.isolate_retcode == 0
+                if isolator.stdout is not None:
+                    self.result['stdout'] = isolator.stdout.decode(
+                        errors='replace'
+                    )
+                if isolator.stderr is not None:
+                    self.result['stderr'] = isolator.stderr.decode(
+                        errors='replace'
+                    )
+                if self.isolator.isolate_stdout is not None:
+                    self.result['isolate'][
+                        'stdout'
+                    ] = isolator.isolate_stdout.decode(errors='replace')
+                if self.isolator.isolate_stderr is not None:
+                    self.result['isolate'][
+                        'stderr'
+                    ] = isolator.isolate_stderr.decode(errors='replace')
+                self.isolator = None
+
+    async def __call__(self, *args, **kwargs):
         try:
-            with log_path.open('r') as f:
-                log = f.read()
-        except FileNotFoundError:
-            pass
-
-    b64compiled = b64encode(compilation_content).decode()
-    return ret, b64compiled, log
-
-
-def isolator_compress(isolator, cmd, path):
-    isolator_path = isolator.path / path
-    try:
-        with isolator_path.open('rb') as fd:
-            return gzip.compress(fd.read())
-    except PermissionError:
-        raise_isolate_error(
-            f"server: {path} does not have the correct permissions.\n",
-            cmd,
-            isolator,
+            async with self.spawn_isolator() as self.isolator:
```

Comment:
Uhhh okay that looks like a weird thing to do to assign to a field though. Isn't `spawn_isolator` already doing this anyway on line 72?